SIL.WritingSystems/Sldr.cs

```diff
@@ -409,49 +410,90 @@ private static void LoadLanguageTagsIfNecessary()
 				CreateSldrCacheDirectory();
 
 				cachedAllTagsPath = Path.Combine(SldrCachePath, "langtags.json");
+				string etagPath;
+				etagPath = Path.Combine(SldrCachePath, "langtags.json.etag");
 				var sinceTime = _embeddedAllTagsTime.ToUniversalTime();
 				if (File.Exists(cachedAllTagsPath))
 				{
 					var fileTime = File.GetLastWriteTimeUtc(cachedAllTagsPath);
 					if (sinceTime > fileTime)
+					{
 						// delete the old langtags.json file if a newer embedded one is available.
 						// this can happen if the application is upgraded to use a newer version of SIL.WritingSystems
 						// that has an updated embedded langtags.json file.
 						File.Delete(cachedAllTagsPath);
+						File.Delete(etagPath);
+					}
 					else
 						sinceTime = fileTime;
 				}
 				sinceTime += TimeSpan.FromSeconds(1);
-				try
-				{
-					if (_offlineMode)
-						throw new WebException("Test mode: SLDR offline so accessing cache", WebExceptionStatus.ConnectFailure);
-
+				
+			}
+			_languageTags = new ReadOnlyKeyedCollection<string, SldrLanguageTagInfo>(ParseAllTagsJson(cachedAllTagsPath));
+		}
 
-					// get SLDR langtags.json from the SLDR api compressed
-					// it will throw WebException or have status HttpStatusCode.NotModified if file is unchanged or not get it
-					var langtagsUrl =
-						$"{SldrRepository}index.html?query=langtags&ext=json{StagingParameter}";
-					var webRequest = (HttpWebRequest) WebRequest.Create(Uri.EscapeUriString(langtagsUrl));
-					webRequest.UserAgent = UserAgent;
-					webRequest.IfModifiedSince = sinceTime;
-					webRequest.Timeout = 10000;
-					webRequest.AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate;
-					using var webResponse = (HttpWebResponse) webRequest.GetResponse();
+		public static void LoadLanguageTags()
```

Comment:
We now have two methods with almost identical names (`LoadLanguageTags` and `LoadLanguageTagsIfNecessary`) which do different things which is confusing. Maybe rename this method to `DownloadLanguageTags`?